### PR TITLE
Fixes #750: Implement Amazon API

### LINKED ISSUE #750: Propose and implement Amazon API
There are product APIs (https://affiliate-program.amazon.com/gp/advertising/api/detail/main.html) and a number of other APIs (https://developer.amazon.com/). Which ones could be relevant for Susi?

* [x] List relevant APIs in this issue
* [ ] As a start implement an API that offers information on products

### PR DESCRIPTION
**Pull request by shivenmian**
@mariobehling @Orbiter 

I have implemented one part of the Amazon Product Advertising API integration, the ItemLookup API. 

Supply an ASIN / Product ID, and if the product is in the database (and if it's approved by it's parent company to be displayed by the API), it gives you necessary info about the result.

For searches like Apple Products (I did a search on iPhone), it isn't available in the Advertising API, because Apple does not want to display the prices of its products that way or something.

If you wish to test, please contact me. I have the AWS Keys and the Associate Tags, which I can't display here right now. I think these fields should be supplied as a GET parameter instead.

TODO:

* [x] Make it more efficient. The Amazon API is a SOAP API, not REST. Some parameters are available in some products, some aren't.

* [x] Have a feature to list other variants of the product, and possibly list the cheapest variant too.

* [x] Add the SimilarityLookup and ItemSearch API (in ItemSearch, you have Search by Product Name).

Screenshot: 

screen shot 2016-08-06 at 5 01 01 pm

### CODE REVIEW

**Comment by sudheesh001** [comment]
Most Amazon products have a table describing the item dimensions etc.., This can be converted to the JSON object too, Can you look into that, It doesn't look like there's a lot of information being collected here.

**Comment by shivenmian** [comment]
@sudheesh001 Yes, the ItemDimensions field was definitely there, I just separated it into Height, Width and Length as you saw.

And yes, there's more to be done, as I mentioned in the TODO.

**Comment by sudheesh001** [comment]
@shivenmian I meant tables like this one
https://www.amazon.com/gp/product/B00OQVZDJM/ref=s9_acss_bw_cg_odsbnc_1b1?pf_rd_m=ATVPDKIKX0DER&pf_rd_s=merchandised-search-2&pf_rd_r=BDM863CM55C13G3866V8&pf_rd_t=101&pf_rd_p=7f7be4a4-f096-4b82-aa90-06a6944c0436&pf_rd_i=6669702011

screen shot 2016-08-06 at 5 38 19 pm

**Comment by shivenmian** [comment]
Oh. Didn't think of this one. Will look into it.

**Comment by shivenmian** [comment]
@mariobehling @sudheesh001 Have completed the Integration.

**API usage:**

Params you need to supply:

1. Amazon Item ID (**id**) or Item Name (**q**). Either of them but one should be there.
2. [Response Group](http://docs.aws.amazon.com/AWSECommerceService/latest/DG/CHAP_ResponseGroupsList.html)(**response_group**). This is optional. If this is not supplied, Response Group is defaulted as Large, and Large gives all the possible info.

File: src/org/loklak/api/amazon/AmazonProductService.java [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+public class AmazonProductService extends AbstractAPIHandler implements APIHandler {
+
+	private static final long serialVersionUID = 2279773523424505716L;
+
+	// keys available on request
+	private static final String AWS_ACCESS_KEY_ID = "";
+	private static final String AWS_SECRET_KEY = "";
+	private static final String ASSOCIATE_TAG = "";
+
```
Comment (sudheesh001) [inline src/org/loklak/api/amazon/AmazonProductService.java]:
```java
	private static final String AWS_ACCESS_KEY_ID = "";
	private static final String AWS_SECRET_KEY = "";
	private static final String ASSOCIATE_TAG = "";
```

These are configurable entries, can you please move them to the `conf` file and use the string fields from there ? We wouldn't expect a user to open a java file and make changes to the KeyID and the Key / Tags

Comment (shivenmian) [inline src/org/loklak/api/amazon/AmazonProductService.java]: Makes sense, hmm.

File: src/org/loklak/api/amazon/AmazonProductService.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+			TransformerFactory tf = TransformerFactory.newInstance();
+			Transformer transformer = tf.newTransformer();
+			transformer.transform(domSource, result);
+			JSONObject xmlresult = new JSONObject(true);
+			xmlresult = XML.toJSONObject(writer.toString());
+			JSONObject items = xmlresult.getJSONObject(operation).getJSONObject("Items");
+			if (items.getJSONObject("Request").has("Errors")) {
+				itemlookup.put("status", "error");
+				itemlookup.put("reason",
+						items.getJSONObject("Request").getJSONObject("Errors").getJSONObject("Error").get("Message"));
```
Comment (sudheesh001) [inline src/org/loklak/api/amazon/AmazonProductService.java]: Will the fetch always be in this format ?

Comment (shivenmian) [inline src/org/loklak/api/amazon/AmazonProductService.java]: Yes. I am implementing the ItemLookup and the ItemSearch API, and they are always in this format.

There are other APIs too: like SimilarityLookup etc, but the result of SimilarityLookup (and all the other APIs) comes when ```response_group=Large```, and as I wrote it's that by default in my code, so if you want all the results, just fetch an ```id``` or a ```q``` and leave response_group blank.

**Pull request merged**